embedded/store/immustore.go

```diff
@@ -875,6 +898,26 @@ func (s *ImmuStore) commit(otx *OngoingTx, expectedHeader *TxHeader, waitForInde
 		return nil, err
 	}
 
+	// first check is performed on the snapshot taken when opening the transaction,
+	// this will prevent acquiring the write lock and putting garbage into
+	// appendables if we can already determine that constraints are not met.
+	//
+	// A corner case is if the DB fails to meet constraints now but would fulfill
+	// those during final commit phase - but in such case, we are allowed to reason
+	// about the DB state in any point in time between both snapshots are checked.
+	if hasConstrainedKVEntries(otx.entries) {
+		snap, err := s.indexer.SnapshotSince(s.indexer.Ts())
```

Comment:
this constraint validation may be done directly over the tbtree. See unsafeIndex struct in this same file

Reply:
Fixed, thanks!